src/Sulu/Bundle/SecurityBundle/Factory/UserFactory.php

```diff
@@ -0,0 +1,152 @@
+
+use Doctrine\ORM\EntityManagerInterface;
+use Sulu\Bundle\ContactBundle\Entity\ContactInterface;
+use Sulu\Bundle\ContactBundle\Entity\ContactRepositoryInterface;
+use Sulu\Bundle\SecurityBundle\Entity\UserRole;
+use Sulu\Component\Localization\Manager\LocalizationManagerInterface;
+use Sulu\Component\Security\Authentication\RoleInterface;
+use Sulu\Component\Security\Authentication\RoleRepositoryInterface;
+use Sulu\Component\Security\Authentication\SaltGenerator;
+use Sulu\Component\Security\Authentication\UserInterface;
+use Sulu\Component\Security\Authentication\UserRepositoryInterface;
+use Symfony\Component\Security\Core\Encoder\EncoderFactoryInterface;
+use Webmozart\Assert\Assert;
+
+final class UserFactory implements UserFactoryInterface
+{
+    /** @var EntityManagerInterface */
+    private $entityManager;
+
+    /** @var UserRepositoryInterface */
+    private $userRepository;
+
+    /** @var RoleRepositoryInterface */
+    private $roleRepository;
+
+    /** @var ContactRepositoryInterface */
+    private $contactRepository;
+
+    /** @var LocalizationManagerInterface */
+    private $localizationManager;
+
+    /** @var SaltGenerator */
+    private $saltGenerator;
+
+    /** @var EncoderFactoryInterface */
+    private $encoderFactory;
+
+    public function __construct(
+        EntityManagerInterface $entityManager,
+        UserRepositoryInterface $userRepository,
+        RoleRepositoryInterface $roleRepository,
+        ContactRepositoryInterface $contactRepository,
+        LocalizationManagerInterface $localizationManager,
+        SaltGenerator $saltGenerator,
+        EncoderFactoryInterface $encoderFactory
+    ) {
+        $this->entityManager = $entityManager;
+        $this->userRepository = $userRepository;
+        $this->roleRepository = $roleRepository;
+        $this->contactRepository = $contactRepository;
+        $this->localizationManager = $localizationManager;
+        $this->saltGenerator = $saltGenerator;
+        $this->encoderFactory = $encoderFactory;
+    }
+
+    public function create(
+        string $username,
+        string $firstName,
+        string $lastName,
+        string $email,
+        string $locale,
+        string $password,
+        string $roleName = 'User'
+    ): UserInterface {
+        $existing = $this->userRepository->findOneBy(['username' => $username]);
+        if ($existing instanceof UserInterface) {
+            return $existing;
+        }
+
+        $locales = \array_keys($this->localizationManager->getLocalizations());
+
+        return $this->createUser($firstName, $lastName, $email, $username, $password, $locale, $roleName, $locales);
+    }
+
+    /**
+     * Generates a random salt for the password.
+     */
+    private function generateSalt(): string
+    {
+        return $this->saltGenerator->getRandomSalt();
+    }
+
+    /**
+     * Encodes the given password, for the given password, with he given salt and returns the result.
```

Comment:
Would also remove that entire comment, and add correct type hints to the function.